`fan_out_request` fails to compile with `result<void>`

Here's a diff that adds a test to current master that I believe should work:

```diff
diff --git a/libcaf_core/test/mixin/requester.cpp b/libcaf_core/test/mixin/requester.cpp
index 380c0c4e1..a43e244f9 100644
--- a/libcaf_core/test/mixin/requester.cpp
+++ b/libcaf_core/test/mixin/requester.cpp
@@ -21,6 +21,7 @@ namespace {
 using discarding_server_type = typed_actor<result<void>(int, int)>;
 
 using adding_server_type = typed_actor<result<int>(int, int)>;
+using no_op_server_type = typed_actor<result<void>(int, int)>;
 
 using result_type = std::variant<none_t, unit_t, int>;
 
@@ -167,6 +168,30 @@ CAF_TEST(requesters support fan_out_request) {
   CHECK_EQ(*sum, 9);
 }
 
+CAF_TEST(requesters support fan_out_request with result<void>) {
+  using policy::select_all;
+  std::vector<no_op_server_type> workers{
+    make_server([](int, int) {}),
+    make_server([](int, int) {}),
+    make_server([](int, int) {}),
+  };
+  run();
+  auto ran = std::make_shared<bool>(false);
+  auto client = sys.spawn([=](event_based_actor* self) {
+    self->fan_out_request<select_all>(workers, infinite, 1, 2).then([=]() {
+      *ran = true;
+    });
+  });
+  run_once();
+  expect((int, int), from(client).to(workers[0]).with(1, 2));
+  expect((), from(workers[0]).to(client));
+  expect((int, int), from(client).to(workers[1]).with(1, 2));
+  expect((), from(workers[1]).to(client));
+  expect((int, int), from(client).to(workers[2]).with(1, 2));
+  expect((), from(workers[2]).to(client));
+  CHECK_EQ(*ran, true);
+}
+
 #ifdef CAF_ENABLE_EXCEPTIONS
 
 CAF_TEST(exceptions while processing requests trigger error messages) {
```

This fails to compile for me using clang-15 on macOS:

```bash
❯ cmake --build build --target libcaf_core/CMakeFiles/caf-core-test.dir/test/mixin/requester.cpp.o
```

<details>
<summary>
Click to expand
</summary>

```
[1/1] Building CXX object libcaf_core/CMakeFiles/caf-core-test.dir/test/mixin/requester.cpp.o
FAILED: libcaf_core/CMakeFiles/caf-core-test.dir/test/mixin/requester.cpp.o
ccache /opt/homebrew/opt/llvm/bin/clang++ -Dlibcaf_core_EXPORTS -Ilibcaf_core/test -Ilibcaf_core -Ibuild/libcaf_core -Ilibcaf_test -isystem /opt/homebrew/opt/llvm/include/c++/v1 -g -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -pedantic -ftemplate-depth=512 -ftemplate-backtrace-limit=0 -Wdocumentation -std=c++17 -MD -MT libcaf_core/CMakeFiles/caf-core-test.dir/test/mixin/requester.cpp.o -MF libcaf_core/CMakeFiles/caf-core-test.dir/test/mixin/requester.cpp.o.d -o libcaf_core/CMakeFiles/caf-core-test.dir/test/mixin/requester.cpp.o -c libcaf_core/test/mixin/requester.cpp
In file included from libcaf_core/test/mixin/requester.cpp:14:
libcaf_core/caf/policy/select_all.hpp:86:1: error: implicit instantiation of undefined template 'caf::detail::select_select_all_helper<(lambda at libcaf_core/test/mixin/requester.cpp:181:69), caf::detail::type_list<>>'
using select_all_helper_t = typename select_select_all_helper<F>::type;
^
libcaf_core/caf/policy/select_all.hpp:106:3: note: in instantiation of template type alias 'select_all_helper_t' requested here
  using type_checker
  ^
libcaf_core/caf/response_handle.hpp:96:27: note: in instantiation of template type alias 'type_checker' requested here
    policy_type::template type_checker<F>::check();
                          ^
libcaf_core/caf/response_handle.hpp:105:5: note: in instantiation of function template specialization 'caf::response_handle<caf::event_based_actor, caf::policy::select_all<caf::detail::type_list<void>>>::then<caf::actor_traits<caf::event_based_actor>, (lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    then(std::move(f), [self](error& err) { self->call_error_handler(err); });
    ^
libcaf_core/test/mixin/requester.cpp:181:64: note: in instantiation of function template specialization 'caf::response_handle<caf::event_based_actor, caf::policy::select_all<caf::detail::type_list<void>>>::then<caf::actor_traits<caf::event_based_actor>, (lambda at libcaf_core/test/mixin/requester.cpp:181:69)>' requested here
    self->fan_out_request<select_all>(workers, infinite, 1, 2).then([=]() {
                                                               ^
libcaf_core/caf/policy/select_all.hpp:72:8: note: template is declared here
struct select_select_all_helper;
       ^
libcaf_core/caf/policy/select_all.hpp:86:1: error: implicit instantiation of undefined template 'caf::detail::select_select_all_helper<(lambda at libcaf_core/test/mixin/requester.cpp:181:69), caf::detail::type_list<>>'
using select_all_helper_t = typename select_select_all_helper<F>::type;
^
libcaf_core/caf/policy/select_all.hpp:168:25: note: in instantiation of template type alias 'select_all_helper_t' requested here
    using helper_type = select_all_helper_t<decay_t<F>>;
                        ^
libcaf_core/caf/policy/select_all.hpp:131:17: note: in instantiation of function template specialization 'caf::policy::select_all<caf::detail::type_list<void>>::make_behavior<(lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    auto bhvr = make_behavior(std::forward<F>(f), std::forward<OnError>(g));
                ^
libcaf_core/caf/response_handle.hpp:97:13: note: in instantiation of function template specialization 'caf::policy::select_all<caf::detail::type_list<void>>::then<caf::event_based_actor, (lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    policy_.then(self_, std::move(f), std::move(g));
            ^
libcaf_core/caf/response_handle.hpp:105:5: note: in instantiation of function template specialization 'caf::response_handle<caf::event_based_actor, caf::policy::select_all<caf::detail::type_list<void>>>::then<caf::actor_traits<caf::event_based_actor>, (lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    then(std::move(f), [self](error& err) { self->call_error_handler(err); });
    ^
libcaf_core/test/mixin/requester.cpp:181:64: note: in instantiation of function template specialization 'caf::response_handle<caf::event_based_actor, caf::policy::select_all<caf::detail::type_list<void>>>::then<caf::actor_traits<caf::event_based_actor>, (lambda at libcaf_core/test/mixin/requester.cpp:181:69)>' requested here
    self->fan_out_request<select_all>(workers, infinite, 1, 2).then([=]() {
                                                               ^
libcaf_core/caf/policy/select_all.hpp:72:8: note: template is declared here
struct select_select_all_helper;
       ^
libcaf_core/caf/policy/select_all.hpp:169:17: error: excess elements in scalar initializer
    helper_type helper{ids_.size(), pending_timeouts_, std::move(f)};
                ^                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libcaf_core/caf/policy/select_all.hpp:131:17: note: in instantiation of function template specialization 'caf::policy::select_all<caf::detail::type_list<void>>::make_behavior<(lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    auto bhvr = make_behavior(std::forward<F>(f), std::forward<OnError>(g));
                ^
libcaf_core/caf/response_handle.hpp:97:13: note: in instantiation of function template specialization 'caf::policy::select_all<caf::detail::type_list<void>>::then<caf::event_based_actor, (lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    policy_.then(self_, std::move(f), std::move(g));
            ^
libcaf_core/caf/response_handle.hpp:105:5: note: in instantiation of function template specialization 'caf::response_handle<caf::event_based_actor, caf::policy::select_all<caf::detail::type_list<void>>>::then<caf::actor_traits<caf::event_based_actor>, (lambda at libcaf_core/test/mixin/requester.cpp:181:69), (lambda at libcaf_core/caf/response_handle.hpp:105:24)>' requested here
    then(std::move(f), [self](error& err) { self->call_error_handler(err); });
    ^
libcaf_core/test/mixin/requester.cpp:181:64: note: in instantiation of function template specialization 'caf::response_handle<caf::event_based_actor, caf::policy::select_all<caf::detail::type_list<void>>>::then<caf::actor_traits<caf::event_based_actor>, (lambda at libcaf_core/test/mixin/requester.cpp:181:69)>' requested here
    self->fan_out_request<select_all>(workers, infinite, 1, 2).then([=]() {
                                                               ^
In file included from libcaf_core/test/mixin/requester.cpp:14:
libcaf_core/caf/policy/select_all.hpp:170:26: error: member reference base type 'helper_type' (aka 'int') is not a structure or union
    auto pending = helper.pending;
                   ~~~~~~^~~~~~~~
In file included from libcaf_core/test/mixin/requester.cpp:9:
In file included from libcaf_core/test/core-test.hpp:9:
In file included from libcaf_test/caf/test/bdd_dsl.hpp:4:
libcaf_test/caf/test/dsl.hpp:280:7: error: no matching function for call to 'extract'
      extract<Ts...>(dest_, src_line_);
      ^~~~~~~~~~~~~~
libcaf_core/test/mixin/requester.cpp:187:3: note: in instantiation of member function 'expect_clause<>::expect_clause' requested here
  expect((), from(workers[0]).to(client));
  ^
libcaf_test/caf/test/dsl.hpp:924:4: note: expanded from macro 'expect'
  (expect_clause<CAF_EXPAND(CAF_DSL_LIST types)>{sched, __LINE__}.fields.eval( \
   ^
libcaf_test/caf/test/dsl.hpp:253:22: note: candidate template ignored: couldn't infer template argument 'T'
std::tuple<T, Ts...> extract(caf_handle x, int src_line) {
                     ^
5 errors generated.
ninja: build stopped: subcommand failed.
```

</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,8 @@ is based on [Keep a Changelog](https://keepachangelog.com).
   on the buffer that could cause indefinite hanging of an application.
 - Fused stages now properly forward errors during the initial subscription to
   their observer.
+- The `fan_out_request` request now properly deals with actor handles that
+  respond with `void` (#1369).
 
 ## [0.19.0-rc.1] - 2022-10-31
 

--- a/libcaf_core/caf/policy/select_all.hpp
+++ b/libcaf_core/caf/policy/select_all.hpp
@@ -37,8 +37,11 @@ using select_all_helper_value_t =
   typename select_all_helper_value_oracle<Ts...>::type;
 
 template <class F, class... Ts>
-struct select_all_helper {
-  using value_type = select_all_helper_value_t<Ts...>;
+struct select_all_helper;
+
+template <class F, class T, class... Ts>
+struct select_all_helper<F, T, Ts...> {
+  using value_type = select_all_helper_value_t<T, Ts...>;
   std::vector<value_type> results;
   std::shared_ptr<size_t> pending;
   disposable timeouts;
@@ -52,10 +55,10 @@ struct select_all_helper {
     results.reserve(pending);
   }
 
-  void operator()(Ts&... xs) {
+  void operator()(T& x, Ts&... xs) {
     CAF_LOG_TRACE(CAF_ARG2("pending", *pending));
     if (*pending > 0) {
-      results.emplace_back(std::move(xs)...);
+      results.emplace_back(std::move(x), std::move(xs)...);
       if (--*pending == 0) {
         timeouts.dispose();
         f(std::move(results));
@@ -64,7 +67,34 @@ struct select_all_helper {
   }
 
   auto wrap() {
-    return [this](Ts&... xs) { (*this)(xs...); };
+    return [this](T& x, Ts&... xs) { (*this)(x, xs...); };
+  }
+};
+
+template <class F>
+struct select_all_helper<F> {
+  std::shared_ptr<size_t> pending;
+  disposable timeouts;
+  F f;
+
+  template <class Fun>
+  select_all_helper(size_t pending, disposable timeouts, Fun&& f)
+    : pending(std::make_shared<size_t>(pending)),
+      timeouts(std::move(timeouts)),
+      f(std::forward<Fun>(f)) {
+    // nop
+  }
+
+  void operator()() {
+    CAF_LOG_TRACE(CAF_ARG2("pending", *pending));
+    if (*pending > 0 && --*pending == 0) {
+      timeouts.dispose();
+      f();
+    }
+  }
+
+  auto wrap() {
+    return [this] { (*this)(); };
   }
 };
 
@@ -80,6 +110,11 @@ struct select_select_all_helper<
 template <class F, class T>
 struct select_select_all_helper<F, detail::type_list<std::vector<T>>> {
   using type = select_all_helper<F, T>;
+};
+
+template <class F>
+struct select_select_all_helper<F, detail::type_list<>> {
+  using type = select_all_helper<F>;
 };
 
 template <class F>

--- a/libcaf_core/test/mixin/requester.cpp
+++ b/libcaf_core/test/mixin/requester.cpp
@@ -21,6 +21,7 @@ namespace {
 using discarding_server_type = typed_actor<result<void>(int, int)>;
 
 using adding_server_type = typed_actor<result<int>(int, int)>;
+using no_op_server_type = typed_actor<result<void>(int, int)>;
 
 using result_type = std::variant<none_t, unit_t, int>;
 
@@ -165,6 +166,30 @@ CAF_TEST(requesters support fan_out_request) {
   expect((int, int), from(client).to(workers[2]).with(1, 2));
   expect((int), from(workers[2]).to(client).with(3));
   CHECK_EQ(*sum, 9);
+}
+
+CAF_TEST(requesters support fan_out_request with result<void>) {
+  using policy::select_all;
+  std::vector<no_op_server_type> workers{
+    make_server([](int, int) {}),
+    make_server([](int, int) {}),
+    make_server([](int, int) {}),
+  };
+  run();
+  auto ran = std::make_shared<bool>(false);
+  auto client = sys.spawn([=](event_based_actor* self) {
+    self->fan_out_request<select_all>(workers, infinite, 1, 2).then([=]() {
+      *ran = true;
+    });
+  });
+  run_once();
+  expect((int, int), from(client).to(workers[0]).with(1, 2));
+  expect((), from(workers[0]).to(client));
+  expect((int, int), from(client).to(workers[1]).with(1, 2));
+  expect((), from(workers[1]).to(client));
+  expect((int, int), from(client).to(workers[2]).with(1, 2));
+  expect((), from(workers[2]).to(client));
+  CHECK_EQ(*ran, true);
 }
 
 #ifdef CAF_ENABLE_EXCEPTIONS

--- a/libcaf_test/caf/test/dsl.hpp
+++ b/libcaf_test/caf/test/dsl.hpp
@@ -249,18 +249,29 @@ std::optional<std::tuple<T, Ts...>> try_extract(caf_handle x) {
 /// Returns the content of the next mailbox element without taking it out of
 /// the mailbox. Fails on an empty mailbox or if the content of the next
 /// element does not match `<T, Ts...>`.
-template <class T, class... Ts>
-std::tuple<T, Ts...> extract(caf_handle x, int src_line) {
-  if (auto result = try_extract<T, Ts...>(x)) {
-    return std::move(*result);
+template <class... Ts>
+std::tuple<Ts...> extract(caf_handle x, int src_line) {
+  if constexpr (sizeof...(Ts) > 0) {
+    if (auto result = try_extract<Ts...>(x)) {
+      return std::move(*result);
+    } else {
+      auto ptr = x->peek_at_next_mailbox_element();
+      if (ptr == nullptr)
+        CAF_FAIL("cannot peek at the next message: mailbox is empty", src_line);
+      else
+        CAF_FAIL("message does not match expected types: "
+                   << to_string(ptr->content()),
+                 src_line);
+    }
   } else {
     auto ptr = x->peek_at_next_mailbox_element();
     if (ptr == nullptr)
       CAF_FAIL("cannot peek at the next message: mailbox is empty", src_line);
-    else
-      CAF_FAIL("message does not match expected types: "
+    else if (!ptr->content().empty())
+      CAF_FAIL("message does not match (expected an empty message): "
                  << to_string(ptr->content()),
                src_line);
+    return {};
   }
 }
 

